src/pkg/cli/tail.go

```diff
@@ -348,81 +362,79 @@ func tail(ctx context.Context, provider client.Provider, projectName string, opt
 					cancel() // TODO: stuck on defer Close() if we don't do this
 					return nil
 				}
-				continue
+				return nil
 			}
 
-			ts := e.Timestamp.AsTime()
-			// Skip duplicate logs (e.g. after reconnecting we might get the same log once more)
-			if skipDuplicate && ts.Equal(options.Since) {
-				skipDuplicate = false
-				continue
-			}
 			if ts.After(options.Since) {
 				options.Since = ts
 			}
-
-			if options.Raw {
-				if e.Stderr {
-					term.Error(e.Message)
-				} else {
-					term.Println(e.Message)
-				}
-				continue
+			err := handler(e, &options)
+			if err != nil {
+				term.Debug("Ending tail loop", err)
+				cancel() // TODO: stuck on defer Close() if we don't do this
 			}
+		}
+	}
+}
 
-			// Replace service progress messages with our own spinner
-			if doSpinner && isProgressDot(e.Message) {
-				continue
-			}
+func TailEntryHandler(e *defangv1.LogEntry, options *TailOptions) error {
```

Comment:
This handler *prints* so I'd make that obvious from the name.

Reply:
Done. By the way, this change set is factored out in https://github.com/DefangLabs/defang/pull/1164 so it can be reviewed and merged separately